A59-audit-logging.md

```diff
@@ -0,0 +1,349 @@
+It is valuable for users to detect anomalous traffic patterns or monitor accesses
+to certain services.
+
+Therefore, authentication and authorization information needs to be available
+in the audit context. Since gRPC authorization is based on the [RBAC policy][RBAC policy],
+we design the audit logging feature around it to provide consistent APIs across
+xDS and non-xDS use cases. The proposed xDS API changes should also benefit
+other xDS clients which are currently lacking this feature as well.
+
+In both xDS and non-xDS cases, we expect users to configure the audit condition
+and audit loggers in language-agnostic configs and have the capability to inject
+their own logger implementations via the APIs we expose. This allows users to
+plug in their own audit logic while not having to restart the servers when
+reconfiguring the loggers.
+
+For auditing purposes, users should not log every incoming RPCs. In other words,
+`ON_DENY_AND_ALLOW` should seldom be used. But this is added as an option to
+support special cases where logging everything is required.
+
+Regarding alternative design options, following are what we have considered in
+a number of othorgonal aspects.
+
+### Utilizing the Access Log
+
+The [information](https://github.com/envoyproxy/envoy/blob/main/api/envoy/data/accesslog/v3/accesslog.proto)
+present in the existing access log is sufficient for audit purposes. However,
+access log does not happen till the response is sent and therefore does not
+fit the audit logging use case well. For example, for long-living RPCs, users
+would want to audit the event right after the authorization is enfoced but
+access log won't necessarily happen till hours later.
+
+Note that there has been ongoing effort in Envoy to support flushing access
+logs at different lifecycle points of a request (see [example](https://github.com/envoyproxy/envoy/pull/26094)).
+It would be a valid approach if the access log is generalized to support what
+is needed here. But gRPC does not and has no plan yet to support access log,
+due to the lack of such need. Therefore, we decided not to take this approach
+which would require more engineering effort.
+
+### Audit Condition in the HTTP Connection Manager
+
+As an HTTP filter chain can contain multiple RBAC filters in general, the typical
+behavior that users will expect if they want to audit on allow is audit should
+only happen once after the evaluation of the last RBAC filter. The gRPC
+authorization policy API assumes this expectation and handles the audit condition
+for the users.
+
+In the xDS cases, however, the audit condition could be considered as something
+on top of all RBAC filters and thus configured in the [HTTP Connection Manager][HttpConnectionManager proto].
+We decided not to take this approach because the component managing the filter
+chain would have to be aware of the last RBAC filter constantly and inform it
```

Comment:
I don't know how to parse "constantly" in this context. Should it just be removed?

Reply:
Removed.